GetBuildVersion fails when git history used version.json for another object

I keep getting following when trying to use the gitversioning package

    1>Target "GetBuildVersion" in file "E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets" from project "E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\src\XLabs.Core\XLabs.Core.csproj" (target "GenerateAssemblyInfo" depends on it):
    1>Using "GetBuildVersion" task from assembly "E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.Tasks.dll".
    1>Task "GetBuildVersion"
    1>E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets(30,5): error MSB4018: The "GetBuildVersion" task failed unexpectedly.
    1>E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets(30,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.
    1>E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets(30,5): error MSB4018:    at Nerdbank.GitVersioning.GitExtensions.GetVersionHeight(Repository repo, String repoRelativeProjectDirectory)
    1>E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets(30,5): error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.Execute()
    1>E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets(30,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
    1>E:\Projects\Github\ravensorb\Xamarin-Forms-Labs\packages\Nerdbank.GitVersioning.1.2.1\build\Nerdbank.GitVersioning.targets(30,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
    1>Done executing task "GetBuildVersion" -- FAILED.
    1>Done building target "GetBuildVersion" in project "XLabs.Core.csproj" -- FAILED.

I have tried a rebuild all, a clean and then build, a full reboot, and still I get the same errors.

Here is the contents of my version.json file (it is at the solution level).  I have also tried at the project level and get the same result

    { "version": "2.0.5818" }

--- a/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs
@@ -82,7 +82,7 @@ public class GitExtensionsTests : RepoTestBase
         // Emulate a repo that used version.json for something else.
         string versionJsonPath = Path.Combine(this.RepoPath, "version.json");
         File.WriteAllText(versionJsonPath, @"{ ""unrelated"": false }");
-        Assert.Equal(1, this.Repo.GetVersionHeight()); // exercise code that handles the file not yet checked in.
+        Assert.Equal(0, this.Repo.GetVersionHeight()); // exercise code that handles the file not yet checked in.
         this.Repo.Stage(versionJsonPath);
         this.Repo.Commit("Add unrelated version.json file.");
         Assert.Equal(1, this.Repo.GetVersionHeight()); // exercise code that handles a checked in file.
@@ -95,8 +95,8 @@ public class GitExtensionsTests : RepoTestBase
 
         // Also emulate case of where the related project.json was just changed to conform,
         // but not yet checked in.
-        this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).Single());
-        Assert.Equal(1, this.Repo.GetVersionHeight());
+        this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).First());
+        Assert.Equal(0, this.Repo.GetVersionHeight());
     }
 
     [Fact]
@@ -108,7 +108,7 @@ public class GitExtensionsTests : RepoTestBase
         // Now introduce a parsing error.
         string versionJsonPath = Path.Combine(this.RepoPath, "version.json");
         File.WriteAllText(versionJsonPath, @"{ ""version"": ""1.0"""); // no closing curly brace for parsing error
-        Assert.Equal(1, this.Repo.GetVersionHeight());
+        Assert.Equal(0, this.Repo.GetVersionHeight());
         this.Repo.Stage(versionJsonPath);
         this.Repo.Commit("Add broken version.json file.");
         Assert.Equal(1, this.Repo.GetVersionHeight());
@@ -118,8 +118,8 @@ public class GitExtensionsTests : RepoTestBase
         Assert.Equal(1, this.Repo.GetVersionHeight());
 
         // And emulate fixing it without having checked in yet.
-        this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).Single());
-        Assert.Equal(1, this.Repo.GetVersionHeight());
+        this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).First());
+        Assert.Equal(0, this.Repo.GetVersionHeight());
     }
 
     [Fact]

--- a/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs
@@ -100,6 +100,29 @@ public class GitExtensionsTests : RepoTestBase
     }
 
     [Fact]
+    public void GetVersionHeight_VersionJsonHasParsingErrorsInHistory()
+    {
+        this.WriteVersionFile();
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+
+        // Now introduce a parsing error.
+        string versionJsonPath = Path.Combine(this.RepoPath, "version.json");
+        File.WriteAllText(versionJsonPath, @"{ ""version"": ""1.0"""); // no closing curly brace for parsing error
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+        this.Repo.Stage(versionJsonPath);
+        this.Repo.Commit("Add broken version.json file.");
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+
+        // Now fix it.
+        this.WriteVersionFile();
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+
+        // And emulate fixing it without having checked in yet.
+        this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).Single());
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+    }
+
+    [Fact]
     public void GetTruncatedCommitIdAsInteger_Roundtrip()
     {
         var firstCommit = this.Repo.Commit("First", this.Signer, this.Signer, new CommitOptions { AllowEmptyCommit = true });

--- a/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs
+++ b/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs
@@ -77,6 +77,29 @@ public class GitExtensionsTests : RepoTestBase
     }
 
     [Fact]
+    public void GetVersionHeight_VersionJsonHasUnrelatedHistory()
+    {
+        // Emulate a repo that used version.json for something else.
+        string versionJsonPath = Path.Combine(this.RepoPath, "version.json");
+        File.WriteAllText(versionJsonPath, @"{ ""unrelated"": false }");
+        Assert.Equal(1, this.Repo.GetVersionHeight()); // exercise code that handles the file not yet checked in.
+        this.Repo.Stage(versionJsonPath);
+        this.Repo.Commit("Add unrelated version.json file.");
+        Assert.Equal(1, this.Repo.GetVersionHeight()); // exercise code that handles a checked in file.
+
+        // And now the repo has decided to use this package.
+        this.WriteVersionFile();
+
+        Assert.Equal(1, this.Repo.Head.GetVersionHeight());
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+
+        // Also emulate case of where the related project.json was just changed to conform,
+        // but not yet checked in.
+        this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).Single());
+        Assert.Equal(1, this.Repo.GetVersionHeight());
+    }
+
+    [Fact]
     public void GetTruncatedCommitIdAsInteger_Roundtrip()
     {
         var firstCommit = this.Repo.Commit("First", this.Signer, this.Signer, new CommitOptions { AllowEmptyCommit = true });

--- a/src/NerdBank.GitVersioning.Tests/project.json
+++ b/src/NerdBank.GitVersioning.Tests/project.json
@@ -5,7 +5,8 @@
     "Nerdbank.GitVersioning": "1.2.1",
     "Newtonsoft.Json": "7.0.1",
     "Validation": "2.0.6.15003",
-    "xunit": "2.1.0"
+    "xunit": "2.1.0",
+    "xunit.runner.visualstudio": "2.1.0"
   },
   "frameworks": {
     ".NETFramework,Version=v4.5.2": { }

--- a/src/NerdBank.GitVersioning/GitExtensions.cs
+++ b/src/NerdBank.GitVersioning/GitExtensions.cs
@@ -63,9 +63,9 @@
             VersionOptions workingCopyVersionOptions, committedVersionOptions;
             if (IsVersionFileChangedInWorkingCopy(repo, repoRelativeProjectDirectory, out committedVersionOptions, out workingCopyVersionOptions))
             {
-                Version workingCopyVersion = workingCopyVersionOptions?.Version.Version;
-                Version headCommitVersion = committedVersionOptions?.Version.Version ?? Version0;
-                if (!workingCopyVersion.Equals(headCommitVersion))
+                Version workingCopyVersion = workingCopyVersionOptions?.Version?.Version;
+                Version headCommitVersion = committedVersionOptions?.Version?.Version ?? Version0;
+                if (workingCopyVersion == null || !workingCopyVersion.Equals(headCommitVersion))
                 {
                     // The working copy has changed the major.minor version.
                     // So by definition the version height is 0, since no commit represents it yet.

--- a/src/NerdBank.GitVersioning/VersionFile.cs
+++ b/src/NerdBank.GitVersioning/VersionFile.cs
@@ -58,7 +58,7 @@
             using (var content = GetVersionFileReader(commit, repoRelativeProjectDirectory, out json))
             {
                 return content != null
-                    ? ReadVersionFile(content, json)
+                    ? TryReadVersionFile(content, json)
                     : null;
             }
         }
@@ -102,7 +102,11 @@
                 {
                     using (var sr = new StreamReader(versionTxtPath))
                     {
-                        return ReadVersionFile(sr, isJsonFile: false);
+                        var result = TryReadVersionFile(sr, isJsonFile: false);
+                        if (result != null)
+                        {
+                            return result;
+                        }
                     }
                 }
 
@@ -111,7 +115,11 @@
                 {
                     using (var sr = new StreamReader(versionJsonPath))
                     {
-                        return ReadVersionFile(sr, isJsonFile: true);
+                        var result = TryReadVersionFile(sr, isJsonFile: true);
+                        if (result != null)
+                        {
+                            return result;
+                        }
                     }
                 }
 
@@ -242,13 +250,20 @@
         /// </summary>
         /// <param name="versionTextContent">The content of the version.txt file to read.</param>
         /// <param name="isJsonFile"><c>true</c> if the file being read is a JSON file; <c>false</c> for the old-style text format.</param>
-        /// <returns>The version information read from the file.</returns>
-        private static VersionOptions ReadVersionFile(TextReader versionTextContent, bool isJsonFile)
+        /// <returns>The version information read from the file; or <c>null</c> if a deserialization error occurs.</returns>
+        private static VersionOptions TryReadVersionFile(TextReader versionTextContent, bool isJsonFile)
         {
             if (isJsonFile)
             {
                 string jsonContent = versionTextContent.ReadToEnd();
-                return JsonConvert.DeserializeObject<VersionOptions>(jsonContent, JsonSettings);
+                try
+                {
+                    return JsonConvert.DeserializeObject<VersionOptions>(jsonContent, JsonSettings);
+                }
+                catch (JsonSerializationException)
+                {
+                    return null;
+                }
             }
 
             string versionLine = versionTextContent.ReadLine();

